### PR TITLE
Update security docs

### LINKED ISSUE #565: update security docs

### PR DESCRIPTION
**Pull request by ericnordelo**
Fixes #565  

#### PR Checklist

- [x] Tests
- [ ] Tried the feature on a public network
- [x] Documentation

### CODE REVIEW

File: src/security/initializable.cairo [hunk 1]
```diff
@@ -1,6 +1,10 @@
 // SPDX-License-Identifier: MIT
 // OpenZeppelin Contracts for Cairo v0.7.0 (security/initializable.cairo)
 
+trait ExternalTrait<TState> {
+    fn is_initialized(self: @TState) -> bool;
+}
```
Comment (andrew-fleming) [inline src/security/initializable.cairo:6]: No strong opinions on including this trait. I suppose there may be cases where this is useful

**Review by andrew-fleming — COMMENTED** [review]
Looking good, sir! I left some comments, questions, and suggestions

File: src/security/pausable.cairo [hunk 2]
```diff
@@ -71,15 +70,15 @@ mod Pausable {
 
         /// Triggers a stopped state.
         /// The contract must not already be paused.
-        fn _pause(ref self: ComponentState<TContractState>) {
+        fn pause(ref self: ComponentState<TContractState>) {
```
[the remark this answers is not in the document]
Comment (ericnordelo) [reply src/security/pausable.cairo:73]: Fair point. Agree is better wait for that discussion to happen.

File: src/security/initializable.cairo (hunk 1 above)
Comment (ericnordelo) [reply src/security/initializable.cairo:6]: The name IPausable sounds a bit misleading IMO (for an interface with just the `is_paused` method), I think for this case is better to use a trait (like PublicKeyTrait in account), I selected the name ExternalTrait because no better name occurred to me for just is_paused. We should document this convention/pattern if we go along with it. What do you think @martriay ?

Comment (andrew-fleming) [reply src/security/initializable.cairo:6]: > The name IPausable sounds a bit misleading IMO (for an interface with just the is_paused method)

Good point. Here, it'd be `IInitializable`, but the point is the same. I can see it being misleading with just the getter

**Review by andrew-fleming — APPROVED** [review]
LGTM!

File: src/tests/security/test_pausable.cairo [hunk 3]
```diff
@@ -1,7 +1,5 @@
-use openzeppelin::security::pausable::PausableComponent::InternalImpl;
-use openzeppelin::security::pausable::PausableComponent::PausableImpl;
-use openzeppelin::security::pausable::PausableComponent::Paused;
-use openzeppelin::security::pausable::PausableComponent::Unpaused;
+use openzeppelin::security::pausable::PausableComponent::{InternalImpl, PausableImpl};
+use openzeppelin::security::pausable::PausableComponent::{Paused, Unpaused};
```
Comment (martriay) [inline src/tests/security/test_pausable.cairo:2]:
```suggestion
use openzeppelin::security::PausableComponent::{InternalImpl, PausableImpl};
use openzeppelin::security::PausableComponent::{Paused, Unpaused};
```
shouldn't we do the same as with the other modules?

File: src/tests/security/test_initializable.cairo [hunk 4]
```diff
@@ -1,4 +1,4 @@
-use openzeppelin::security::initializable::InitializableComponent::InternalImpl;
+use openzeppelin::security::InitializableComponent::{InitializableImpl, InternalImpl};
```
Comment (martriay) [inline src/tests/security/test_initializable.cairo:1]: why was this needed?

File: src/tests/mocks/initializable_mock.cairo [hunk 5]
```diff
@@ -4,6 +4,9 @@ mod InitializableMock {
 
     component!(path: InitializableComponent, storage: initializable, event: InitializableEvent);
 
+    #[abi(embed_v0)]
+    impl InitializableImpl =
+        InitializableComponent::InitializableImpl<ContractState>;
```
Comment (martriay) [inline src/tests/mocks/initializable_mock.cairo:9]: why do we need this?

File: src/security/initializable.cairo (hunk 1 above)
Comment (martriay) [reply src/security/initializable.cairo:6]: > We should document this convention/pattern if we go along with it.

I just [wrote this](https://github.com/OpenZeppelin/cairo-contracts/issues/696#issuecomment-1803235090) recently in our style guide, which is something I recall we discussed maybe during a call. Let me know if you think it doesn't apply here

> The name IPausable sounds a bit misleading IMO (for an interface with just the `is_paused` method)

I kindof see what you mean, but on the other hand "pausable" doesn't really mean I can pause it, but rather that it can be paused, so the getter is not really confusing to me. btw Contracts for Solidity's version has [public pause/unpause methods](https://github.com/ConsenSysMesh/openzeppelin-solidity/blob/master/contracts/lifecycle/Pausable.sol#L37-L48)

File: src/security/pausable.cairo [hunk 6]
```diff
@@ -1,8 +1,7 @@
 // SPDX-License-Identifier: MIT
 // OpenZeppelin Contracts for Cairo v0.8.0-beta.0 (security/pausable.cairo)
 
-#[starknet::interface]
-trait IPausable<TState> {
+trait ExternalTrait<TState> {
```
Comment (martriay) [inline src/security/pausable.cairo:4]: I think `IPausable` is fine

File: docs/modules/ROOT/pages/security.adoc [hunk 7]
```diff
@@ -1,121 +1,193 @@
 = Security
 
-The following documentation provides context, reasoning, and examples of methods and constants found in `openzeppelin/security/`.
+The following documentation provides context, reasoning, and examples of modules found in `openzeppelin/security/`.
```
Comment (martriay) [inline docs/modules/ROOT/pages/security.adoc:3]:
```suggestion
The following documentation provides context, reasoning, and examples of modules found under `openzeppelin::security`.
```

File: docs/modules/ROOT/pages/security.adoc [hunk 8]
```diff
@@ -1,121 +1,193 @@
...
- ** <<safeuint256,SafeUint256>>
+CAUTION: Expect these modules to evolve.
 
 == Initializable
 
-The Initializable library provides a simple mechanism that mimics the functionality of a constructor.
-More specifically, it enables logic to be performed once and only once which is useful to setup a contract's initial state when a constructor cannot be used.
+The xref:api/security.adoc#InitializableComponent[Initializable] component provides a simple mechanism that mimics
+the functionality of a constructor.
+More specifically, it enables logic to be performed once and only once which is useful to set up a contract's initial state when a constructor cannot be used, for example, when there are circular dependencies at construction time.
```
Comment (martriay) [inline docs/modules/ROOT/pages/security.adoc:11]:
```suggestion
More specifically, it enables logic to be performed once and only once which is useful to set up a contract's initial state when a constructor cannot be used, for example when there are circular dependencies at construction time.
```

File: docs/modules/ROOT/pages/security.adoc [hunk 9]
```diff
@@ -1,121 +1,193 @@
...
+            // Do something
+
+            self.reentrancy_guard.end();
+        }
+    }
 }
 ----
+
+NOTE: The guard prevents the execution flow occurring inside `protected_function` to call `another_protected_function`,
+and vice versa.
```
Comment (martriay) [inline docs/modules/ROOT/pages/security.adoc:193]:
```suggestion
NOTE: The guard prevents the execution flow occurring inside `protected_function`
to call itself, to `another_protected_function`, and vice versa.
```

File: docs/modules/ROOT/pages/api/security.adoc [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+= Security
+
+Reference of components, interfaces and utilities found in the library's `security/` directory.
+
+== Initializable
+
+[.contract]
+[[InitializableComponent]]
+=== `++InitializableComponent++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-v0.7.0/src/security/initializable.cairo[{github-icon},role=heading-link]
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/security.adoc:11]:
```suggestion
=== `++InitializableComponent++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-v0.8.0-beta.0/src/security/initializable.cairo[{github-icon},role=heading-link]
```

File: docs/modules/ROOT/pages/api/security.adoc [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+
+Requirements:
+
+- the contract must not have been initialized before.
+
+== Pausable
+
+[.contract]
+[[PausableComponent]]
+=== `++PausableComponent++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-v0.7.0/src/security/pausable.cairo[{github-icon},role=heading-link]
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/security.adoc:61]:
```suggestion
=== `++PausableComponent++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-v0.8.0-beta.0/src/security/pausable.cairo[{github-icon},role=heading-link]
```

File: docs/modules/ROOT/pages/api/security.adoc [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+[[PausableComponent-Unpaused]]
+==== `[.contract-item-name]#++Unpaused++#++(account: ContractAddress)++` [.item-kind]#event#
+
+Emitted when the pause is lifted by `account`.
+
+== ReentrancyGuard
+
+[.contract]
+[[ReentrancyGuardComponent]]
+=== `++ReentrancyGuardComponent++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-v0.7.0/src/security/reentrancyguard.cairo[{github-icon},role=heading-link]
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/security.adoc:166]:
```suggestion
=== `++ReentrancyGuardComponent++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-v0.8.0-beta.0/src/security/reentrancyguard.cairo[{github-icon},role=heading-link]
```

File: docs/modules/ROOT/pages/api/security.adoc [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+Returns whether the contract has been initialized.
+
+[#InitializableComponent-Internal-Functions]
+==== Internal Functions
+
+[.contract-item]
+[[InitializableComponent-initialize]]
+==== `[.contract-item-name]#++initialize++#++(ref self: ComponentState)++` [.item-kind]#internal#
+
+Ensures the calling function can only be called once.
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/security.adoc:51]:
```suggestion
Initializes the contract. Can only be called once.
```

File: docs/modules/ROOT/pages/api/security.adoc [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+--
+
+[#PausableComponent-Embeddable-Functions]
+==== Embeddable Functions
+
+[.contract-item]
+[[PausableComponent-is_paused]]
+==== `[.contract-item-name]#++is_paused++#++(self: @ComponentState)++ → bool` [.item-kind]#external#
+
+Returns whether the contract is currently on pause.
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/security.adoc:105]:
```suggestion
Returns whether the contract is currently paused.
```

File: docs/modules/ROOT/pages/api/security.adoc [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+
+[#PausableComponent-Internal-Functions]
+==== Internal Functions
+
+[.contract-item]
+[[PausableComponent-assert_not_paused]]
+==== `[.contract-item-name]#++assert_not_paused++#++(self: @ComponentState)++` [.item-kind]#internal#
+
+Makes a function only callable when the contract is not paused.
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/security.adoc:114]:
```suggestion
Panics if the contract is paused.
```
I think this is more explicit and clear than "not callable", which might simply mean it does nothing. It's also important to be clear when a function reverts execution.

File: docs/modules/ROOT/pages/api/security.adoc [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+
+[.contract-item]
+[[PausableComponent-assert_paused]]
+==== `[.contract-item-name]#++assert_paused++#++(self: @ComponentState)++` [.item-kind]#internal#
+
+Makes a function only callable when the contract is paused.
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/security.adoc:120]:
```suggestion
Panics if the contract is not paused.
```

File: docs/modules/ROOT/pages/api/security.adoc [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+[.contract-item]
+[[PausableComponent-_pause]]
+==== `[.contract-item-name]#++_pause++#++(ref self: ComponentState)++` [.item-kind]#internal#
+
+Pauses the contract.
+
+Requirements:
+
+- the contract must not be paused.
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/security.adoc:130]:
```suggestion
Requires the contract not to be paused.
```

I'm not sure how I feel about these single element lists, maybe we should be flexible with this format.

File: docs/modules/ROOT/pages/api/security.adoc [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+
+[#PausableComponent-Events]
+==== Events
+
+[.contract-item]
+[[PausableComponent-Paused]]
+==== `[.contract-item-name]#++Paused++#++(account: ContractAddress)++` [.item-kind]#event#
+
+Emitted when the pause is triggered by `account`.
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/security.adoc:154]:
```suggestion
Emitted when the contract is paused by `account`.
```
I think this reads better

File: docs/modules/ROOT/pages/api/security.adoc [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+
+[.contract-item]
+[[PausableComponent-Unpaused]]
+==== `[.contract-item-name]#++Unpaused++#++(account: ContractAddress)++` [.item-kind]#event#
+
+Emitted when the pause is lifted by `account`.
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/security.adoc:160]:
```suggestion
Emitted when the contract is unpaused by `account`.
```

File: docs/modules/ROOT/pages/api/security.adoc [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+
+[.contract]
+[[ReentrancyGuardComponent]]
+=== `++ReentrancyGuardComponent++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-v0.7.0/src/security/reentrancyguard.cairo[{github-icon},role=heading-link]
+
+```javascript
+use openzeppelin::security::ReentrancyGuardComponent;
+```
+
+Component to help prevent nested (reentrant) calls.
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/security.adoc:172]:
```suggestion
Component to help prevent reentrant calls.
```

File: docs/modules/ROOT/pages/api/security.adoc [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+
+[.contract-item]
+[[ReentrancyGuardComponent-start]]
+==== `[.contract-item-name]#++start++#++(ref self: ComponentState)++` [.item-kind]#internal#
+
+Prevents a contract's function from calling itself or another protected function, directly or indirectly.
+
+Requirements:
+
+- the contract must not be currently in a guarded state.
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/security.adoc:194]:
```suggestion
Requires the `ReentrancyGuard` to not be currently enabled.
```
what do you think?

**Review by martriay — COMMENTED** [review]
Docs look really good! Left some comments.

File: src/tests/security/test_initializable.cairo (hunk 4 above)
Comment (ericnordelo) [reply src/tests/security/test_initializable.cairo:1]: To call `is_initialized` later

File: src/tests/mocks/initializable_mock.cairo (hunk 5 above)
Comment (ericnordelo) [reply src/tests/mocks/initializable_mock.cairo:9]: We don't

File: docs/modules/ROOT/pages/api/security.adoc (hunk 17 above)
Comment (ericnordelo) [reply docs/modules/ROOT/pages/api/security.adoc:130]: Maybe is better to address that in a separate issue? There are other places where this is used as well, and I would like to maintain the consistency

File: docs/modules/ROOT/pages/api/security.adoc (hunk 21 above)
Comment (ericnordelo) [reply docs/modules/ROOT/pages/api/security.adoc:194]: The sentence read better, but as [mentioned before](https://github.com/OpenZeppelin/cairo-contracts/pull/794#discussion_r1389317656), I think the single line requirement could be a separate issue and PR.

File: src/tests/security/test_initializable.cairo (hunk 4 above)
Comment (martriay) [reply src/tests/security/test_initializable.cairo:1]: but why wasn't it failing?

File: src/tests/mocks/initializable_mock.cairo [hunk 22]
```diff
@@ -7,7 +7,6 @@ mod InitializableMock {
     #[abi(embed_v0)]
     impl InitializableImpl =
         InitializableComponent::InitializableImpl<ContractState>;
-    impl InternalImpl = InitializableComponent::InternalImpl<ContractState>;
```
Comment (martriay) [inline src/tests/mocks/initializable_mock.cairo:7]: to be honest i don't fully understand why `initialize()` is available for the initializable mock, since the internal impl is not instantiated nor embedded. Is it because the internal impl is being imported in the test file?

**Review by martriay — APPROVED** [review]

File: src/tests/mocks/initializable_mock.cairo (hunk 22 above)
Comment (ericnordelo) [reply src/tests/mocks/initializable_mock.cairo:7]: Yes, it works because there is an implementation in context
